Some suggestions on the residentKey

We developed Server using simpleWebAuthn(5.4.5) library. Test by FIDO Conformance Tools 1.7.2

Currently, SimpleWebAuthn is importing the requireResidentKey using the value of the residentKey, as shown in the example below.

![무제-페이지-21](https://user-images.githubusercontent.com/57163578/185028474-eab5237e-c396-4e5e-b62b-f15f0f1b5d5a.jpeg)

This is in line with Level 2 recommendations.
https://www.w3.org/TR/webauthn/#dictdef-authenticatorselectioncriteria

However, we are trying to get Level1 certification. So I don't use the residentKey.

So, try to get the residentKey, the value comes in undefined.

I thought like the next picture to solve this problem.

![무제-페이지-2](https://user-images.githubusercontent.com/57163578/185028562-d28487e3-85b2-4328-9805-f272c33e28bb.jpeg)

I thought that the compatibility with Level1 would be guaranteed to some extent if it was solved like the picture.

Can you give some suggestions to solve this?

> There is no problem with the test. I just have a question about the code, so I'd like to discuss it.

Ah, I understand now, thanks for the extra context.

Re-reading the definitions of `residentKey` and `requireResidentKey` in the L2 spec, I think your interpretation is probably the best way to accommodate both the sentence you highlighted...

> residentKey, of type DOMString
>
>    Specifies the extent to which the Relying Party desires to create a client-side discoverable credential. For historical reasons the naming retains the deprecated “resident” terminology. The value SHOULD be a member of ResidentKeyRequirement but client platforms MUST ignore unknown values, treating an unknown value as if the member does not exist. **If no value is given then the effective value is required if requireResidentKey is true or discouraged if it is false or absent.**

...and the bit in `requireResidentKey` about defaulting to false but being set to true when `residentKey` is required:
     
> requireResidentKey, of type boolean, **defaulting to false**
>
>    This member is retained for backwards compatibility with WebAuthn Level 1 and, for historical reasons, its naming retains the deprecated “resident” terminology for discoverable credentials. **Relying Parties SHOULD set it to true if, and only if, residentKey is set to required.**

(Emphasis mine)

> In conclusion, I thought about changing the current code like this.
> 
> ![image](https://user-images.githubusercontent.com/57163578/185066967-ed47093b-ac37-4457-82d8-0a72374f1a28.png)
> 
> I wanted to suggest that if you change this part to something like this, it will be a more flexible code that passes through both Level 2 and Level 1.

Thank you for taking the time to think up alternative logic for `requireResidentKey` and `residentKey`. I was satisfied enough with your explanation that I created #259 to incorporate these changes.

Note that I wasn't able to initialize `residentKey` to `"discouraged"` when absent and `requireResidentKey` is `false` because when I did so FIDO Conformance v1.7.2 failed the very first test:

```
AssertionError: Response.authenticatorSelection MUST be set to the requested authenticatorSelection!

Expected
"{"requireResidentKey":false,"userVerification":"preferred","residentKey":"discouraged"}"
to equal
"{"requireResidentKey":false,"userVerification":"preferred"}":
```

If I left out just that one branch of logic then FIDO Conformance tests remain passing.

--- a/packages/server/src/registration/generateRegistrationOptions.test.ts
+++ b/packages/server/src/registration/generateRegistrationOptions.test.ts
@@ -49,6 +49,7 @@ test('should generate credential request options suitable for sending via JSON',
     excludeCredentials: [],
     authenticatorSelection: {
       requireResidentKey: false,
+      residentKey: 'discouraged',
       userVerification: 'preferred',
     },
   });
@@ -120,6 +121,7 @@ test('should set authenticatorSelection if specified', () => {
   expect(options.authenticatorSelection).toEqual({
     authenticatorAttachment: 'cross-platform',
     requireResidentKey: false,
+    residentKey: 'discouraged',
     userVerification: 'preferred',
   });
 });
@@ -165,4 +167,91 @@ test('should use custom supported algorithm IDs as-is when provided', () => {
     { alg: -8, type: 'public-key' },
     { alg: -65535, type: 'public-key' },
   ]);
+});
+
+test('should require resident key if residentKey option is absent but requireResidentKey is set to true', () => {
+  const options = generateRegistrationOptions({
+    rpID: 'not.real',
+    rpName: 'SimpleWebAuthn',
+    userID: '1234',
+    userName: 'usernameHere',
+    authenticatorSelection: {
+      requireResidentKey: true,
+    }
+  });
+
+  expect(options.authenticatorSelection?.requireResidentKey).toEqual(true);
+  expect(options.authenticatorSelection?.residentKey).toEqual('required');
+});
+
+test('should discourage resident key if residentKey option is absent but requireResidentKey is set to false', () => {
+  const options = generateRegistrationOptions({
+    rpID: 'not.real',
+    rpName: 'SimpleWebAuthn',
+    userID: '1234',
+    userName: 'usernameHere',
+    authenticatorSelection: {
+      requireResidentKey: false,
+    }
+  });
+
+  expect(options.authenticatorSelection?.requireResidentKey).toEqual(false);
+  expect(options.authenticatorSelection?.residentKey).toEqual('discouraged');
+});
+
+test('should discourage resident key if both residentKey and requireResidentKey options are absent', () => {
+  const options = generateRegistrationOptions({
+    rpID: 'not.real',
+    rpName: 'SimpleWebAuthn',
+    userID: '1234',
+    userName: 'usernameHere',
+  });
+
+  expect(options.authenticatorSelection?.requireResidentKey).toEqual(false);
+  expect(options.authenticatorSelection?.residentKey).toEqual('discouraged');
+});
+
+test('should set requireResidentKey to true if residentKey if set to required', () => {
+  const options = generateRegistrationOptions({
+    rpID: 'not.real',
+    rpName: 'SimpleWebAuthn',
+    userID: '1234',
+    userName: 'usernameHere',
+    authenticatorSelection: {
+      residentKey: 'required',
+    },
+  });
+
+  expect(options.authenticatorSelection?.requireResidentKey).toEqual(true);
+  expect(options.authenticatorSelection?.residentKey).toEqual('required');
+});
+
+test('should set requireResidentKey to false if residentKey if set to preferred', () => {
+  const options = generateRegistrationOptions({
+    rpID: 'not.real',
+    rpName: 'SimpleWebAuthn',
+    userID: '1234',
+    userName: 'usernameHere',
+    authenticatorSelection: {
+      residentKey: 'preferred',
+    },
+  });
+
+  expect(options.authenticatorSelection?.requireResidentKey).toEqual(false);
+  expect(options.authenticatorSelection?.residentKey).toEqual('preferred');
+});
+
+test('should set requireResidentKey to false if residentKey if set to discouraged', () => {
+  const options = generateRegistrationOptions({
+    rpID: 'not.real',
+    rpName: 'SimpleWebAuthn',
+    userID: '1234',
+    userName: 'usernameHere',
+    authenticatorSelection: {
+      residentKey: 'discouraged',
+    },
+  });
+
+  expect(options.authenticatorSelection?.requireResidentKey).toEqual(false);
+  expect(options.authenticatorSelection?.residentKey).toEqual('discouraged');
 });

--- a/packages/server/src/registration/generateRegistrationOptions.test.ts
+++ b/packages/server/src/registration/generateRegistrationOptions.test.ts
@@ -49,7 +49,6 @@ test('should generate credential request options suitable for sending via JSON',
     excludeCredentials: [],
     authenticatorSelection: {
       requireResidentKey: false,
-      residentKey: 'discouraged',
       userVerification: 'preferred',
     },
   });
@@ -121,7 +120,6 @@ test('should set authenticatorSelection if specified', () => {
   expect(options.authenticatorSelection).toEqual({
     authenticatorAttachment: 'cross-platform',
     requireResidentKey: false,
-    residentKey: 'discouraged',
     userVerification: 'preferred',
   });
 });
@@ -196,10 +194,10 @@ test('should discourage resident key if residentKey option is absent but require
   });
 
   expect(options.authenticatorSelection?.requireResidentKey).toEqual(false);
-  expect(options.authenticatorSelection?.residentKey).toEqual('discouraged');
+  expect(options.authenticatorSelection?.residentKey).toBeUndefined();
 });
 
-test('should discourage resident key if both residentKey and requireResidentKey options are absent', () => {
+test('should not set resident key if both residentKey and requireResidentKey options are absent', () => {
   const options = generateRegistrationOptions({
     rpID: 'not.real',
     rpName: 'SimpleWebAuthn',
@@ -208,7 +206,7 @@ test('should discourage resident key if both residentKey and requireResidentKey 
   });
 
   expect(options.authenticatorSelection?.requireResidentKey).toEqual(false);
-  expect(options.authenticatorSelection?.residentKey).toEqual('discouraged');
+  expect(options.authenticatorSelection?.residentKey).toBeUndefined();
 });
 
 test('should set requireResidentKey to true if residentKey if set to required', () => {

--- a/packages/server/src/registration/generateRegistrationOptions.ts
+++ b/packages/server/src/registration/generateRegistrationOptions.ts
@@ -120,15 +120,30 @@ export function generateRegistrationOptions(
   }));
 
   /**
-   * "Relying Parties SHOULD set [requireResidentKey] to true if, and only if, residentKey is set
-   * to "required""
    *
-   * See https://www.w3.org/TR/webauthn-2/#dom-authenticatorselectioncriteria-requireresidentkey
    */
-  if (authenticatorSelection.residentKey === 'required') {
-    authenticatorSelection.requireResidentKey = true;
+  if (authenticatorSelection.residentKey === undefined) {
+    /**
+     * `residentKey`: "If no value is given then the effective value is `required` if
+     * requireResidentKey is true or `discouraged` if it is false or absent."
+     *
+     * See https://www.w3.org/TR/webauthn-2/#dom-authenticatorselectioncriteria-residentkey
+     */
+    if (authenticatorSelection.requireResidentKey) {
+      authenticatorSelection.residentKey = 'required';
+    } else {
+      authenticatorSelection.residentKey = 'discouraged';
+    }
   } else {
-    authenticatorSelection.requireResidentKey = false;
+    /**
+     * `requireResidentKey`: "Relying Parties SHOULD set it to true if, and only if, residentKey is
+     * set to "required""
+     *
+     * Spec says this property defaults to `false` so we should still be okay to assign `false` too
+     *
+     * See https://www.w3.org/TR/webauthn-2/#dom-authenticatorselectioncriteria-requireresidentkey
+     */
+    authenticatorSelection.requireResidentKey = authenticatorSelection.residentKey === 'required';
   }
 
   return {

--- a/packages/server/src/registration/generateRegistrationOptions.ts
+++ b/packages/server/src/registration/generateRegistrationOptions.ts
@@ -120,7 +120,8 @@ export function generateRegistrationOptions(
   }));
 
   /**
-   *
+   * Capture some of the nuances of how `residentKey` and `requireResidentKey` how either is set
+   * depending on when either is defined in the options
    */
   if (authenticatorSelection.residentKey === undefined) {
     /**
@@ -132,7 +133,11 @@ export function generateRegistrationOptions(
     if (authenticatorSelection.requireResidentKey) {
       authenticatorSelection.residentKey = 'required';
     } else {
-      authenticatorSelection.residentKey = 'discouraged';
+      /**
+       * FIDO Conformance v1.7.2 fails the first test if we do this, even though this is
+       * technically compatible with the WebAuthn L2 spec...
+       */
+      // authenticatorSelection.residentKey = 'discouraged';
     }
   } else {
     /**

